grpc/server.go

```diff
@@ -169,17 +191,84 @@ func (sb *serverBuilder) Build(tlsConfig *tls.Config, statsRegistry prometheus.R
 		return server.Serve(listener)
 	}
 
+	// Initialize long-running health checks of all services which implement the
+	// checker interface.
+	if sb.checkInterval <= 0 {
+		sb.checkInterval = 5 * time.Second
+	}
+	healthCtx, stopHealthChecks := context.WithCancel(context.Background())
+	for _, s := range sb.services {
+		check, ok := s.impl.(checker)
+		if !ok {
+			continue
+		}
+		sb.initLongRunningCheck(healthCtx, s.desc.ServiceName, check.Health)
+	}
+
 	// Start a goroutine which listens for a termination signal, and then
 	// gracefully stops the gRPC server. This in turn causes the start() function
 	// to exit, allowing its caller (generally a main() function) to exit.
 	go cmd.CatchSignals(func() {
-		healthSrv.Shutdown()
+		stopHealthChecks()
+		sb.healthSrv.Shutdown()
 		server.GracefulStop()
 	})
 
 	return start, nil
 }
 
+// initLongRunningCheck initializes a goroutine which will periodically check
+// the health of the provided service and update the health server accordingly.
+func (sb *serverBuilder) initLongRunningCheck(healthCtx context.Context, service string, checkImpl func(context.Context) error) {
+	go func() {
+		ticker := time.NewTicker(sb.checkInterval)
+		defer ticker.Stop()
+
+		// Set the initial health status for the service.
+		sb.healthSrv.SetServingStatus(service, healthpb.HealthCheckResponse_NOT_SERVING)
+		lastStatus := healthpb.HealthCheckResponse_NOT_SERVING
+
+		// check is a helper function which performs the health check for the
+		// service.
+		check := func() (healthpb.HealthCheckResponse_ServingStatus, error) {
```

Comment:
This function can be defined outside this `go func` (the one that starts on line 223), to reduce nesting and make understanding the actual goroutine easier.

Also, it is currently a closure over healthCtx, but it doesn't need to be an imo shouldn't be: it should be a function which takes a context.Context as its argument, and lines 260 and 266 should pass healthCtx as that argument.